srcjs/init_shiny.js

```diff
@@ -332,13 +338,49 @@ function initShiny() {
   }
 
   $('.shiny-image-output, .shiny-plot-output, .shiny-report-theme').each(function() {
-    var id = getIdFromEl(this);
-    initialValues['.clientdata_output_' + id + '_bg'] = getComputedBgColor(this);
-    initialValues['.clientdata_output_' + id + '_fg'] = getStyle(this, "color");
-    initialValues['.clientdata_output_' + id + '_accent'] = getComputedLinkColor(this);
-    initialValues['.clientdata_output_' + id + '_font'] = getComputedFont(this);
+    var el = this, id = getIdFromEl(el);
+    initialValues['.clientdata_output_' + id + '_bg'] = getComputedBgColor(el);
+    initialValues['.clientdata_output_' + id + '_fg'] = getStyle(el, "color");
+    initialValues['.clientdata_output_' + id + '_accent'] = getComputedLinkColor(el);
+    initialValues['.clientdata_output_' + id + '_font'] = getComputedFont(el);
+    maybeAddThemeObserver(el);
   });
 
+
+  // Resend computed styles if this element's class or style changes which is
+  // especially important for invalidating getCurrentOutputInfo() when error
+  // messages modify the containers CSS in a non-welcome way.
+  // https://github.com/rstudio/shiny/issues/3196
+  //
+  // Unfortunately, won't solve getCurrentOutputInfo() not invalidating
+  // if this element's ancestors change in a meaningful way, but I don't
+  // see a reasonable way to solve that issue
+  // https://github.com/rstudio/shiny/issues/2998
+  function maybeAddThemeObserver(el) {
+    var $el = $(el);
+    if ($el.data("shiny-theme-observer")) return;
+
+    var cl = el.classList;
+    var reportTheme = cl.contains('shiny-image-output') || cl.contains('shiny-plot-output') || cl.contains('shiny-report-theme');
+    if (!reportTheme) return;
+
+    var observerCallback = new Debouncer(null, () => doSendTheme(el), 100);
+    var observer = new MutationObserver(() => observerCallback.normalCall());
+    var config = {attributes: true, attributeFilter: ['style', 'class']};
+    observer.observe(el, config);
+    $el.data("shiny-theme-observer", true); // TODO: remove this flag on unbind?
```

Comment:
Instead of setting to `true`, this could be set to `observer`, which would also provide an interface for disconnecting the observer.